[Q]: Step15-1

park78951 opened the pull request:
미션 15-1 검토 부탁드립니다.

디버깅에서 굉장히 오래 시간을 소비한 것 같습니다... 디버깅을 하다보니 욕심이나서 더 고치고 수정하고, 하다보니 또 버그가 발견되고... 코드의 순서에 따른 버그 발생에 대한 관계를 이해하려다 보니 시간이 더 거렸던 것 같습니다. 최대한 리팩토링을 시도했지만 여전히 마음에 들지는 않습니다...  
지금은 html을 캐러셀과 나눴지만 다음 미션에서 합치도록 하겠습니다.

1. 기능 구현

• 검색어 입력 후 'Enter' 입력 또는 검색 아이콘 'Click'시 데이터 저장 후 창 종료
• 검색어 입력시 자동완성 창 생성
   • 해당 검색어가 들어간 단어에 대해 내림차순 정렬된 리스트 생성
   • 입력 후 300ms 이후 자동완성 창 생성
   • 해당 검색어에 해당하는 부분 굵은 글씨 표시
   • 최대 5개의 리스트 생성
• 앞서 저장된 데이터를 바탕으로 최근검색 리스트 생성
   • 가장 최근 검색된 검색어가 맨 위에 출력
   • 중복되는 검색어 입력시 가장 위에 오도록 설정
   • 최대 5개의 리스트 생성
• input 창 외에 클릭시 창의 Focus를 없애고 출력되었던 리스트 (최근검색 or 자동완성) 숨김
• 다시 input창 클릭시 리스트(최근검색 or 자동완성) 출력

2. 문제점 인식 및 의문점
• 초반 기능 구현에 초점을 맞추고 기능 구현 시 버그 수정을 하도록 초기 설계하였다가 디버깅에 너무 큰 어려움을 겪어 미리 예외처리 및 세밀한 설계가 필요함을 느꼈습니다만.. 초기 예외처리를 모두 파악하기가 어렵다면 추후 버그 발견시 최대한 쉽게 수정할 수 있는 방법으로 사전에 어떻게 작업을 해야할지 궁금합니다...
• 디자인 패턴을 최대한 사용하려다보니 검색창의 view는 필요성은 기능이 거의 없어 지워버렸고, 자동완성과 최근검색의 view 공통부분은 main view를 만들어 관리하였는데 호출시 main view에서 자동완성이나 최근검색의 메서드를 호출하면서 man view 클래스 자체의 메서드도 호출하다보니 코드가 굉장히 복잡해지고 어떤 방식이 옳은 것인지에 대해 궁금합니다...

3. 미구현 기능
• Fetch data - 자동완성을 위한 데이터들은 단순 무작위 단어 배열로 처리

park78951 committed:
add blocking event function when animating

PR 머지 이후 Branch 변경 및 reabse 작업 진행.
carousel.js 파일에 animation 기능 수행 중 추가 animation 일어나지 않도록
버특 막는 기능 추가

park78951 committed:
add blocking event function when animating

PR 머지 이후 Branch 변경 및 reabse 작업 진행.
carousel.js 파일에 animation 기능 수행 중 추가 animation 일어나지 않도록
버특 막는 기능 추가

park78951 committed:
create promise.js file

Trying to Promise-like codes to implment like Promise library to
understand Promise.
Create MyPromise class and test successful and failure async codes
and successful sync codes.

park78951 committed:
create promise.js file

Trying to Promise-like codes to implment like Promise library to
understand Promise.
Create MyPromise class and test successful and failure async codes
and successful sync codes.

park78951 committed:
create cards.json file

create json file for cards information based on index.html.

park78951 committed:
create cards.json file

create json file for cards information based on index.html.

park78951 committed:
create index.html file and change the file names

create index.html file to implement fetchJson.js with localData.json to
fetch data from json file and change name of files ( cars.json ->
localData.json, fetch.js -> fetchJson.js )

park78951 committed:
Create 2 files for JSON and data model

create data.js file to fetch data from localData.json and set is as a
state in the Data class. With this, create localData.json file to store
information about image URLs, titles and description of contents in
carousel cards.

park78951 committed:
Create 2 files for JSON and data model

create data.js file to fetch data from localData.json and set is as a
state in the Data class. With this, create localData.json file to store
information about image URLs, titles and description of contents in
carousel cards.

park78951 committed:
Create viewer.js file

create viewer.js file to render on the page by using json file fetched
by data.js. It's not yet tested and needs to set dependencies with
other moudles or classes.

park78951 committed:
Create viewer.js file

create viewer.js file to render on the page by using json file fetched
by data.js. It's not yet tested and needs to set dependencies with
other moudles or classes.

park78951 committed:
Create localData.json and viewer.js and modifiy app.js

create viewer.js file to render caraousel main contents using template literal and modify app.js to fetch data from localData.json asynchronously and call viewer.render method with viewer class instantiated.

park78951 committed:
Create localData.json and viewer.js and modifiy app.js

create viewer.js file to render caraousel main contents using template literal and modify app.js to fetch data from localData.json asynchronously and call viewer.render method with viewer class instantiated.

park78951 committed:
Modify viewer.js and app.js, create template.js

Modify viewer.js file to get a part of header(navigator) rendered with
json file by Viewer.js and create viewer.js file only to make html
templates.

A part of change from carousel.js is to get class name 'active' for
hover in style sheet due to the reason of new render of the header items and
the class name active should be assigned for one among them.

park78951 committed:
Modify viewer.js and app.js, create template.js

Modify viewer.js file to get a part of header(navigator) rendered with
json file by Viewer.js and create viewer.js file only to make html
templates.

A part of change from carousel.js is to get class name 'active' for
hover in style sheet due to the reason of new render of the header items and
the class name active should be assigned for one among them.

park78951 committed:
Rename URL of fetch api in app.js

park78951 committed:
Rename URL of fetch api in app.js

park78951 committed:
Organize codes

park78951 committed:
Modify mypromise.js of condition in resolve and reject

park78951 committed:
Merge remote-tracking branch 'upstream/park78951' into step14-1

# Conflicts:
#	index.html
#	mypromise/myPromise.js
#	src/css/style.css
#	src/js/carousel.js
#	src/scss/style.css.map
#	src/scss/style.scss

park78951 committed:
Figure out conflict after merge with step13-1

park78951 committed:
Refactor codes in app.js and rename

1. Refactor codes to be clean: reduce 'then' chain to render mainViewer
in app.js
2. Rename function 'getData' clearly to 'getCarouselData' in app.js
3. Rename file and class name: Viewer.js and Viewer -> RenderTemplate
both

park78951 committed:
Refactor codes in app.js and rename

1. Refactor codes to be clean: reduce 'then' chain to render mainViewer
in app.js
2. Rename function 'getData' clearly to 'getCarouselData' in app.js
3. Rename file and class name: Viewer.js and Viewer -> RenderTemplate
both

park78951 committed:
Add dataset in header 'li' items with method deleted

Add id dataset in order in header items of teplate literal and change
carousel.js file with one of methods, getting an index of card lists,
deleted and use a property to get dataset id to get card index.
(need to get index 0 to 3 and set id from 1 to 4 in each list in order)

park78951 committed:
Add dataset in header 'li' items with method deleted

Add id dataset in order in header items of teplate literal and change
carousel.js file with one of methods, getting an index of card lists,
deleted and use a property to get dataset id to get card index.
(need to get index 0 to 3 and set id from 1 to 4 in each list in order)

park78951 committed:
Add event delegation at buttons

Add button container '<div>' tag in index.html and event delegation at the container in carousel.js

park78951 committed:
Add event delegation at buttons

Add button container '<div>' tag in index.html and event delegation at the container in carousel.js

park78951 committed:
Add event delegation at a parent node of header

Add event delegation at 'carousel__header' class (ul tag) and make only
header items deliver its information among the clicked targets to get
data attributes.

park78951 committed:
Separate folders with files created.

Create files and sepearate them with right folders after first design
for step15-1

park78951 committed:
Separate folders with files created.

Create files and sepearate them with right folders after first design
for step15-1

park78951 committed:
Create methods and dom to store an input value.

1. Create a dom of input tag
2. Create a method to get value from input box in searchBarView.js
3. Create a method store the value in searchModel.js
4. Create a method for eventlistener to check if input event occured and
get event object to get the value.

park78951 committed:
Add Autocomplete function

1. add methods in AutocompleteView.js
(to attach ul tag and li tags with matched words in it,
dettach li tag with the words in the ul tag)
2. add a method in searchModel.js
(to get matched words in autocomplete datas)
3. add methods controlSearch.js
(to store input datas in autocomData in searchModel.js,
to show list for autocomplete datas)

park78951 committed:
Add Autocomplete function

1. add methods in AutocompleteView.js
(to attach ul tag and li tags with matched words in it,
dettach li tag with the words in the ul tag)
2. add a method in searchModel.js
(to get matched words in autocomplete datas)
3. add methods controlSearch.js
(to store input datas in autocomData in searchModel.js,
to show list for autocomplete datas)

park78951 committed:
Add TODO for this project

park78951 committed:
Add autocomplete function with bold matched letter

1. Change to add ul tag in html from making its template when
autocomplete view instantiated.
2. Add simple style for the search bars and autocomplete list.
3. Add methods to make matched letters bold and colored in
AutocompleteView.js and literal template for '<b>' tag in Template.js
4. Change some methods by adding params or changing methods for that
function mentioned in 3.

park78951 committed:
Add autocomplete function with bold matched letter

1. Change to add ul tag in html from making its template when
autocomplete view instantiated.
2. Add simple style for the search bars and autocomplete list.
3. Add methods to make matched letters bold and colored in
AutocompleteView.js and literal template for '<b>' tag in Template.js
4. Change some methods by adding params or changing methods for that
function mentioned in 3.

park78951 committed:
Add function to select list in autocomplete list by key press.

1. add mehthod in autocompleteView.js
   - add properties to store changed stutus.
   - add focusItem method to manage all selection view in autocomplete list.
   - add firstFocus method to select at first when nothing has not been
   selected.
   - add moveFocus method to move selection to another in list.
   - add focusToggle method to add or remove class to highlight on one
   of list
   - add resetFocus and clearAutocomList methods to be reset the
   properties to select list.
2. add methods in controlSearch.js
   - add setTimeout in removeAutocomList method to delete autocomplete
   list after made with setTimeout in order.
   - add selectAutocomList to controll selecting list.
      - use preventdefault to prevent cursor in input focused from being
      moved with kepressed by up and down arrows.
   - addEventHanler for the selectAutocomList method.
3. add ul tag in html to replace template literal and style in css for
the function to select autocomplete list.

park78951 committed:
Add function to select list in autocomplete list by key press.

1. add mehthod in autocompleteView.js
   - add properties to store changed stutus.
   - add focusItem method to manage all selection view in autocomplete list.
   - add firstFocus method to select at first when nothing has not been
   selected.
   - add moveFocus method to move selection to another in list.
   - add focusToggle method to add or remove class to highlight on one
   of list
   - add resetFocus and clearAutocomList methods to be reset the
   properties to select list.
2. add methods in controlSearch.js
   - add setTimeout in removeAutocomList method to delete autocomplete
   list after made with setTimeout in order.
   - add selectAutocomList to controll selecting list.
      - use preventdefault to prevent cursor in input focused from being
      moved with kepressed by up and down arrows.
   - addEventHanler for the selectAutocomList method.
3. add ul tag in html to replace template literal and style in css for
the function to select autocomplete list.

park78951 committed:
Restructure files and codes

1. Create MainView.js file as a mainview of autocompleteView.js and
recentSearchView.js
2. Restructure mark-up features: make 2 container for autocomplete and
recent search and control involved lists by hiding container(recent
search, not removed once made) and remaking list (autocomplete, new list
made whenever typed)
3. Add cleartimeout to fix bug not to be deleted when all input
removed.

park78951 committed:
Restructure files and codes

1. Create MainView.js file as a mainview of autocompleteView.js and
recentSearchView.js
2. Restructure mark-up features: make 2 container for autocomplete and
recent search and control involved lists by hiding container(recent
search, not removed once made) and remaking list (autocomplete, new list
made whenever typed)
3. Add cleartimeout to fix bug not to be deleted when all input
removed.

park78951 committed:
Activate focus function

1. Reuse old methods before restructuring to focus autocomplete list by
highlighting.
2. Add a function to change text in search bar to another text focused
in autocomplete list when highligted.

park78951 committed:
Activate focus function

1. Reuse old methods before restructuring to focus autocomplete list by
highlighting.
2. Add a function to change text in search bar to another text focused
in autocomplete list when highligted.

park78951 committed:
Add a function to focus list of auto view in recent search

1. Make recent search list focused by clicking arrow keys up and down.
2. Bug fixed
   - fix set time-out event problem: make 2 overlapped lists from recent search and autocomplete because of setTimeout rendered in order by using clear timeout.
   - fix unfocusing problem: make recent search list focused after
   deleting all in search bar.

park78951 committed:
Add a function to focus list of auto view in recent search

1. Make recent search list focused by clicking arrow keys up and down.
2. Bug fixed
   - fix set time-out event problem: make 2 overlapped lists from recent search and autocomplete because of setTimeout rendered in order by using clear timeout.
   - fix unfocusing problem: make recent search list focused after
   deleting all in search bar.

park78951 committed:
Reorganize codes to be clean

park78951 committed:
Reorganize codes to be clean

park78951 committed:
Add gitignore config

park78951 committed:
Removed node_module folder

park78951 committed:
Bug fixed

1. Fix a problem not to create recent search list - Have recent search list created newly every time new letters input.

park78951 committed:
Bug fixed

1. Fix a problem not to create recent search list - Have recent search list created newly every time new letters input.

park78951 committed:
Separate class names and reorganize codes

park78951 committed:
Separate class names and reorganize codes

park78951 committed:
Resolve conflicts on merge

park78951 committed:
Git pull origin to resolve conflicts

park78951 committed:
add comment in gitignore to remove

park78951 committed:
add comment in gitignore to remove

crongro commented:
몇차례 말씀드렸지만, 
'디자인 패턴을 최대한 사용하려다보니 ' 라는 부분이  전 틀렸다고 봅니다.
MVC등을 참고는 할 수 있지만 내 상황에 맞는 걸 만들려고 노력해보는 게 좋아요.

- 변경되는부분 분리하기
- 의존성을 낮게 유지하기

라는 점을 중요하게 생각하면서요.

crongro commented on src/js/search/model/searchModel.js:
```diff
@@ -0,0 +1,45 @@
+export default class searchModel {
+  constructor(limitedNum) {
+    this.recentSearches = [];
+    this.autocomData = [
```
autocomData, 
이부분은 정적인데이터들이니까 모델이 있지 않고 config파일등으로 분리해두는 게 좋을 거 같네요.

park78951 replied:
model에서 검색 관련 데이터를 관리하는게 좋아보여서 넣었었는데 말씀하신대로
앞으로 정적인 데이터는 따로 관리하도록 하는게 좋을 것 같습니다.

crongro commented on src/js/search/view/AutocompleteView.js:
```diff
@@ -0,0 +1,67 @@
+
+  render(words, inputVal) {
+    const autocomList = this.attachAutocomList(words, inputVal);
+    ut.appendElLastly(this.autocomplete, autocomList);
```
appendElLastly 함수이름이 잘 안읽히는거 같네요.

crongro reviewed (changes requested):
search/ 부분의 model/view/control 구분은 괜찮아보이는데요.

지금상태가  역할이 잘 나눠져 있으니, 이상태에서 버그를 모두 해결하는 것도 전 좋을거 같고요.
또는, 
단순화 시켜서 다시 구현하려면,  controller를 제거하는 것도 좋을 거 같아요. 그렇게 되면 view에서 model을 바로 구독하는 것이죠.

**어떤 방법으로 개선하던지 버그를 최소화 시켜서 만들어보세요!**

마지막으로, 전체적으로 디렉토리 구조가 복잡한데요. (carousel , promise등등) 
정리하면 좋겠네요.

park78951 committed:
Rework on feedback

1. Reorder directory path.
2. Bug fixed.
3. Relocate static data for autocomplete UI to config.
4. Rename a method "appendElLastly" on myUtil to appendAtLast.

park78951 commented:
피드백 감사합니다~ 
피드백 주신 점 반영해서 수정했습니다.

이번 미션에서 공통적인 역할을 Main View로 뺀 점이 굉장히 구좋를 복잡하게 만들지 않을까 생각했는데, 크롱이 말씀해주신 것과 다른 멤버 코드 보면서 단순하게 구현하기 위해 상속하도록 수정하고 controller도 없애보면서 어떤점이 불편한지, 더 나은지 판단해보면 좋을 것 같네요.

현재 Amazon Mission을 완료한 다음에 싹 다시 짜보도록 하겠습니다.

버그 있었던 점도 수정했습니다.
[데모](https://park78951.github.io/codesquad-Q/)

crongro reviewed (approved):
별 다른 수정없이 버그가 해결된거 같군요.
수정내용 확인했어요.
리뷰가 늦어서 미안합니다 흑
